pallets/gear/src/manager.rs

```diff
@@ -0,0 +1,264 @@
+            persistent_pages.insert(key.raw(), value.to_vec());
+        }
+
+        let id = H256::from_slice(program.id().as_slice());
+
+        let code_hash: H256 = sp_io::hashing::blake2_256(program.code()).into();
+
+        common::set_code(code_hash, program.code());
+
+        let program = common::Program {
+            static_pages: program.static_pages(),
+            nonce: program.message_nonce(),
+            persistent_pages: persistent_pages.keys().copied().collect(),
+            code_hash,
+        };
+
+        common::set_program(id, program, persistent_pages);
+    }
+}
+
+impl<T> JournalHandler for ExtManager<T>
+where
+    T: Config,
+    T::AccountId: Origin,
+{
+    fn execution_fail(
+        &mut self,
+        origin: MessageId,
+        initiator: ProgramId,
+        program_id: ProgramId,
+        reason: &'static str,
+        entry: DispatchKind,
+    ) {
+        let origin = H256::from_slice(origin.as_slice());
+        let program_id = H256::from_slice(program_id.as_slice());
+
+        if let DispatchKind::Init = entry {
+            ProgramsLimbo::<T>::insert(program_id, origin);
+            log::info!(
+                target: "runtime::gear",
+                "👻 Program {} will stay in limbo until explicitly removed",
+                program_id
+            );
+
+            Pallet::<T>::deposit_event(Event::InitFailure(
+                MessageInfo {
+                    message_id: origin,
+                    program_id,
+                    origin: H256::from_slice(initiator.as_slice()),
+                },
+                Reason::Dispatch(reason.as_bytes().to_vec()),
+            ));
+        } else {
+            if let common::value_tree::ConsumeResult::RefundExternal(external, gas_left) = self
+                .gas_tree
+                .as_mut()
+                .expect("Can't fail")
+                .clone()
+                .consume()
+            {
+                let refund = T::GasConverter::gas_to_fee(gas_left);
+
+                let _ = T::Currency::unreserve(
+                    &<T::AccountId as Origin>::from_origin(external),
+                    refund,
+                );
+            }
+
+            Pallet::<T>::deposit_event(Event::MessageDispatched(DispatchOutcome {
+                message_id: origin,
+                outcome: ExecutionResult::Failure(reason.as_bytes().to_vec()),
+            }));
+        }
+    }
+    fn gas_burned(&mut self, origin: MessageId, amount: u64) {
+        // Adjust block gas allowance
+        GasAllowance::<T>::mutate(|x| *x = x.saturating_sub(amount));
+
+        // TODO: weight to fee calculator might not be identity fee
+        let charge = T::GasConverter::gas_to_fee(amount);
+
+        self.gas_tree.as_mut().expect("Can't fail").spend(amount);
+
+        if let Err(e) = T::Currency::transfer(
+            &<T::AccountId as Origin>::from_origin(H256::from_slice(origin.as_slice())),
```

Comment:
Is it supposed to charge fee from a message?

Reply:
I guess the method should accept `origin: ProgramId` & `message_id: MessageId` so the message initiator could be charged